generator/transit_generator.cpp

```diff
@@ -43,20 +61,76 @@ string GetFileName(string const & filePath)
   return name;
 }
 
-/// \brief Reads from |root| (json) and serializes an array to |serializer|.
-/// \param handler is function which fixes up vector of |Item|(s) after deserialization from json
-/// but before serialization to mwm.
 template <class Item>
-void SerializeObject(my::Json const & root, string const & key, Serializer<FileWriter> & serializer,
-                     function<void(vector<Item> &)> handler = nullptr)
+void DeserializeFromJson(my::Json const & root, string const & key, vector<Item> & items)
 {
-  vector<Item> items;
-
+  items.clear();
   DeserializerFromJson deserializer(root.get());
   deserializer(items, key.c_str());
+}
+
+void DeserializeGatesFromJson(my::Json const & root, string const & mwmDir, string const & countryId,
+                              vector<Gate> & gates)
+{
+  DeserializeFromJson(root, "gates", gates);
+
+  // Creating IndexRouter.
+  SingleMwmIndex singleIndex(my::JoinFoldersToPath(mwmDir, countryId + DATA_FILE_EXTENSION));
 
-  if (handler)
-    handler(items);
+  auto infoGetter = storage::CountryInfoReader::CreateCountryInfoReader(GetPlatform());
+  CHECK(infoGetter, ());
+
+  auto const countryFileGetter = [&infoGetter](m2::PointD const & pt) {
+    return infoGetter->GetRegionCountryId(pt);
+  };
+
+  auto const getMwmRectByName = [&](string const & c) -> m2::RectD {
+    CHECK_EQUAL(countryId, c, ());
+    return infoGetter->GetLimitRectForLeaf(c);
+  };
+
+  CHECK_EQUAL(singleIndex.GetMwmId().GetInfo()->GetType(), MwmInfo::COUNTRY, ());
+  auto numMwmIds = make_shared<NumMwmIds>();
+  numMwmIds->RegisterFile(CountryFile(countryId));
+
+  // Note. |indexRouter| is valid until |index| is valid.
+  IndexRouter indexRouter(VehicleType::Pedestrian, false /* load altitudes */,
+                          CountryParentNameGetterFn(), countryFileGetter, getMwmRectByName,
+                          numMwmIds, MakeNumMwmTree(*numMwmIds, *infoGetter),
+                          traffic::TrafficCache(), singleIndex.GetIndex());
+
+  // Looking for the best segment for every gate.
+  for (auto & gate : gates)
+  {
+    // Note. For pedestrian routing all the segments are considered as twoway segments so
+    // IndexRouter.FindBestSegment() method finds the same segment for |isOutgoing| == true
+    // and |isOutgoing| == false.
+    Segment bestSegment;
+    try
+    {
+      if (indexRouter.FindBestSegmentAtSingleMwm(gate.GetPoint(),
+                                                 m2::PointD::Zero() /* direction */,
+                                                 true /* isOutgoing */, bestSegment))
+      {
+        CHECK_EQUAL(bestSegment.GetMwmId(), 0, ());
+        gate.SetBestPedestrianSegment(SingleMwmSegment(
+            bestSegment.GetFeatureId(), bestSegment.GetSegmentIdx(), bestSegment.IsForward()));
+      }
+    }
+    catch (RootException const & e)
+    {
+      LOG(LDEBUG, ("Point of a gate belongs to several mwm or doesn't belong any mwm. gate:", gate,
```

Comment:
s/several mwm/several mwms/
s/belong any/belong to any/

Reply:
Fixed

Reply:
> s/gate:/Gate:/

Fixed.